nemo/collections/nlp/modules/common/transformer/text_generable_model.py

```diff
@@ -0,0 +1,59 @@
+from typing import List, TypedDict, Union
+
+from torch import Tensor
+
+
+class SamplingParam(TypedDict):
+    max_length: int  # The maximum length of the sequence to be generated.
+    min_length: int  # The minimum length of the sequence to be generated.
+    do_sample: bool  # Whether or not to use sampling ; use greedy decoding otherwise
+    top_k: int  # The number of highest probability vocabulary tokens to keep for top-k-filtering.
+    top_p: float  # If set to float < 1, only the most probable tokens with probabilities that add up to top_p or higher are kept for generation.
+    repetition_penalty: float  # The parameter for repetition penalty. 1.0 means no penalty.
+    add_BOS: bool  # add the bos token at the begining of the prompt
+
+
+class OutputType(TypedDict):
+    sentences: List[str]  # output sentences
+    tokens: List[List[str]]  # output sentences borken into tokens
+    logits: List[Tensor]  # logits output for all the tokens
+    token_ids: List[Tensor]  # output sentence token ids
+
+
+class TextGenerableModelInterface:
+    """
+    Interface for all models that is able to generate text.
+
+    """
+
+    def generate(self, inputs: Union[List[str], Tensor, List[dict]], sampling_params: SamplingParam) -> OutputType:
+        """
+        Public method for all models that is able to generate text.
+        @params
+            inputs: Union[List[str], Tensor, List[dict]]
+                Can be one of the 3 types: 
+                1. List of strings. Each element of the list provides input prompt. The model will apply tokenizer on it.
+                    E.g [‘sentence’, ‘sentence2’ … ]
+                2. Pytorch Tensor of shape (batch_size, seq_length).  The sequence used as a prompt for the generation or as model inputs to the encoder. It will skip the tokenization step.
+                    E.g. torch.tensor([[23,5234,23,35…], [223,323,23,23232,232] …])    
+                3. List of python dict objects. Used for prompt/p-tuning inputs where a set of key-value pairs are converted into input token embeddings for the model.
```

Comment:
In prompt/p-tuning case are we only going to support strings? What about tokenized and/or tensor modes?

Reply:
For Prompt/p-tuning model, once the model is trained, the embeddings for the virtual tokens are fixed. We can add virtual tokens to the vocab and expand the embedding table to include those virtual token embeddings which are generated from the prompt encoder.  This is the step that is needed for preparing the model for fastertransformer so it has the best compatibility with GPT models.

Once we finished the virtual tokens vocab expansion step, we can simply use the Tensor type token id input to generate text.

Reply:
@vadam5  what do you think?